strawberryfields/backends/bosonicbackend/bosoniccircuit.py

```diff
@@ -803,6 +804,59 @@ def measure_dyne(self, covmat, modes, shots=1):
 
         return vals
 
+    def measure_fock(self, modes, shots=1):
+        mu = self.means
+        mean = self.get_mean_xp()
+        cov = self.get_covmat_xp()
+        weights = self.weights
+        
+        x_idxs = np.array(modes)
+        p_idxs = (x_idxs + len(np.transpose(mu))/2).astype(int)
+        modes_idxs = np.concatenate([x_idxs, p_idxs])
+        measurements = np.zeros((len(mu)))
+
+        ##measure photon number in every gaussian mode
+        for i in range(len(cov)):
+            cov_i = cov[i]
+            mean_i = mean[i]
+            reduced_cov_i = cov_i[np.ix_(modes_idxs, modes_idxs)]
+            reduced_mean_i = mean_i[modes_idxs]
+            if np.allclose(mu, np.zeros_like(mu)):
+                haf_i = hafnian_sample_state(reduced_cov_i, shots)
+                measurements[i] = haf_i[0][0] 
+            else:
+                haf_i = hafnian_sample_state(reduced_cov_i, shots, mean=reduced_mean_i)
+                measurements[i] = haf_i[0][0] 
+
+        index = np.array([np.random.choice(list(range(len(measurements))),1,p=np.abs(weights))])
+        samples = measurements[index] 
+        return samples
+    
+    def measure_threshold(self, modes, shots=1):
+        mu = self.means
+        cov = self.get_covmat_xp()
+        weights = self.weights
+        
+        x_idxs = np.array(modes)
+        p_idxs = (x_idxs + len(np.transpose(mu))/2).astype(int)
+        modes_idxs = np.concatenate([x_idxs, p_idxs])
+        measurements = np.zeros((len(mu)))
+        
+        ##measure photon number in every gaussian mode
+        for i in range(len(cov)):
+            cov_i = cov[i]
+            reduced_cov_i = cov_i[np.ix_(modes_idxs, modes_idxs)]
+            tor_i = torontonian_sample_state(reduced_cov_i, shots)
+            measurements[i] = tor_i[0][0] 
+
+        index = np.array([np.random.choice(list(range(len(measurements))),1,p=np.abs(weights))])
+        samples = measurements[index]
+        if measurements[index] == 1:
+            print("wip : updating the state in the case there is a click")
+        else:
+            print("wip : updating the state in the case there is no click")     
+        return samples
+
```

Comment:
dear @JeyKelly : my first suggestion would be to use the method [`fidelity_vacuum`](https://github.com/XanaduAI/strawberryfields/blob/9f6debbb22476bebb7ef6d4c3041e3590aeba1d4/strawberryfields/backends/bosonicbackend/bosoniccircuit.py#L542). This method will tell you the probability `p0` that the detector does not click. The probability that it does click is given by `1-p0`. You can then flip a coin with this bias (using `np.random.choice`) and then based on that one will have to update the vectors of coefficients, means, and covariance matrices. Maybe I'd suggest you start by writing the logic deciding whether to click or not. I will try to give you more detailed info on how to update the state based on the two different outcomes.

Reply:
@nquesada This is definitely a more efficient solution, thank you. It works properly except in the case of vacuum where I used  `is_vacuum` to circumvent the rounding error (`p0>1`). Is there a similarly simple solution for the Fock measurement or is sampling each gaussian mode individually like I put in the pull request a decent approach? It still has the issue of using non-normalized weights, so in the case of a pure Fock state, a measurement can't be done using the current definition.

Reply:
Hi @JeyKelly --- As promised here are some notes explaining the algorithm:

[bosonic_threshold.pdf](https://github.com/XanaduAI/strawberryfields/files/6764815/bosonic_threshold.pdf)

Reply:
Also, this update rule was implemented in Fortran in this repo here https://github.com/XanaduAI/torontonian-sampling/blob/master/src/torontonian_samples.f90

Reply:
> 
> 
> Hi @JeyKelly --- As promised here are some notes explaining the algorithm:
> 
> [bosonic_threshold.pdf](https://github.com/XanaduAI/strawberryfields/files/6764815/bosonic_threshold.pdf)

I took a look at this, thanks for explaining it. I was wondering though: wouldn't the rewritten cov matrix \sigma^{(l)} have (2M - 2) for matrices dimensions excluding the last mode in \sigma_{(B)} (and similarly for \vec{r}_{(A)})? I'm asking because I have a hard time understanding why the cov matrices have a (M - 2) x (M - 2) sizes post no-click measurements. 

The way I interpret this is we're cutting out the mode completely post-measurement, and so we end up with modified matrices for the remaining states - however, I don't quite understand why in the case where we have a click, the remaining cov/weights/means matrices have sizes which reflect this extra mode that we have measured? Initially, I thought we were replacing the mode with vacuum and adjusting the remaining ones to reflect this depending on the measurement outcome, but as far as I understand it's not what is done here. Do you have an idea where I'm not understanding properly?

Reply:
> Also, this update rule was implemented in Fortran in this repo here https://github.com/XanaduAI/torontonian-sampling/blob/master/src/torontonian_samples.f90

@nquesada I am having trouble interfacing this in Python; it's telling me I'm missing the dependencies (use kinds, vars, structures) in the `torontonian_samples` module. Would this ultimately be translated in python code to be implemented in Strawberryfields? If I can manage to make it run on my end I would gladly rewrite it in a python file that can be used directly in the measure_threshold method.

Reply:
Hi @jeykelly --- I am not suggesting you use that Fortran code verbatim. I suggested it because it can give you and idea of how to implement the update rules. Also, I made typo, in the notes, the updated matrices are of size 2M-2 x 2M-2. Sorry about this!

Reply:
also to answer your question @JeyKelly about why they are 2M-2 x 2M-2, it is because they only describe the modes which were not measured. Since optical measurements are typically _demolition_ measurements, the measured mode is typically reset to vacuum, as you said. This can be done afterwards, just be adding 2 extra rows/cols to the covariance matrix in the positions corresponding to the measured modes (just with hbar/2 on the diagonal and zero everywhere else). And adding 2 extra zeros to the updated 2M-2 length displacement vector.